docs: add identity wallet deployment guide (#524)
Mobility-Data-Space/mds-edc: docs/deployment/wallet_deployment.md

For the accompanying **credential issuer** service (which issues Membership / DataProcessor credentials to wallets), see the [mds-identity-issuer](https://github.com/Mobility-Data-Space/mds-identity-issuer) repository.

## When you need a wallet

You need one wallet **per connector participant** as soon as you are running the DCP (Decentralized Claims Protocol) profile. Legacy DAPS-based connectors do not need it. The DCP launchers (`connector-vault-postgresql-dcp`, `connector-inmemory-dcp`) rely on a wallet reachable at `edc.iam.sts.oauth.token.url`, so the wallet must be running and the participant context registered before the connector will complete a Dataspace Protocol handshake.

## Prerequisites

- **PostgreSQL** — dedicated database (or dedicated schema on a shared instance). The wallet owns its own tables via Flyway (`extensions/identity-hub/database-schema-migration-wallet`); do **not** share the schema with the connector.
- **HashiCorp Vault** — for private key storage. The wallet stores generated participant private keys under `<did>-privatekey-alias`, per-participant super-user api-keys, and the STS client secrets.

> - **A resolvable `did:web` hostname** — the wallet publishes the DID document at `https://<wallet-host>/.well-known/did.json`; whichever DIDs your participants use (`did:web:<connector-host>`) must point at *this* wallet's `/` for resolution to work.

Reviewer: Can we be a bit more clear about this? 
My understanding is that the DID:Web **is** the IdentityHub Wallet resolvable did.json. 
It is **not** the connector host.
When I read this, I understand that you are recommending we point the connector-host to the wallet-host (which would work). I am not sure this is recommended.

My suggestion is to simply have a wallet-host and a connector-host, they are separate and refer to 2 different systems.

Author: Agreed — reworked the doc around this model: the DID is now `did:web:<wallet-host>`, the wallet serves its own did.json, and the connector host is a separate system throughout. The ingress-forwarding pattern is gone.

- **The issuer's DID resolvable** (needed later for credential requests) — either at a public URL or reachable from the wallet pod.

## Building the wallet image

The wallet launcher is not currently published to `ghcr.io`. Build it locally from source:

```bash
# from the repo root
./gradlew :launchers:wallet:build
docker build \
  --build-arg RUNTIME=wallet \
  -t mds-wallet:local \
  -f launchers/Dockerfile launchers/
```

> The `Dockerfile` at `launchers/Dockerfile` is shared across all runtimes and picks the launcher via the `RUNTIME` build-arg (same pattern as `connector-vault-postgresql`). The resulting image runs `bin/wallet` as its entrypoint.

Reviewer: The image will be built and provided through Github. I think we should remove this.

Author: Done — removed the build section and replaced it with a pointer to the published image. Adding the wallet runtime to `build-push.yml` will be done in another PR.

EDC_IDENTITYHUB_SUPERUSER_ID=super-user
EDC_IDENTITYHUB_SUPERUSER_API_KEY=<base64(super-user)>.<random>

# did:web must use HTTPS in production so counterparties can resolve DIDs.
EDC_IAM_DID_WEB_USE_HTTPS=true
```

The super-user api-key must be formatted `base64(<EDC_IDENTITYHUB_SUPERUSER_ID>).<random-string>` — a random-only value boots but every subsequent admin call returns 401 with the log line `this key appears to have an invalid format`.

The seeder writes the api-key into Vault under `secret/<super-user-id>-apikey` on first boot **only**; if the participant-context row already exists in Postgres, the vault write is skipped. If you rotate the k8s secret without wiping the DB row you will get 401s until you either delete the row + Vault entry and restart, or write the new value into Vault by hand.

## Registering a participant (self-provide a wallet)

Once the wallet is up you register **one participant context per connector** through the admin Identity API. The context is what the connector's STS calls will authenticate against.

The example below assumes:

- `WALLET_HOST=https://wallet.example.com` — the wallet's public hostname.

> - `PARTICIPANT_DID=did:web:my-connector.example.com` — the DID your connector will advertise on DSP.

Reviewer: ```suggestion
- `WALLET_HOST=https://wallet.example.com` — the wallet's public hostname.
- `PARTICIPANT_DID=did:web:wallet.example.com` — the DID your connector will advertise on DSP.
```

This is related to the above point as well. Let me know if this make sense to you

Author: Applied — the example now uses `did:web:wallet.example.com` (folded into the wallet-host rework above, wording slightly adapted).

- `SUPER_API_KEY` — the value of `EDC_IDENTITYHUB_SUPERUSER_API_KEY`.

```bash
PARTICIPANT_CONTEXT_ID="$(uuidgen)"

curl -fsS -X POST "${WALLET_HOST}/api/identity/v1beta/participants" \
  -H "x-api-key: ${SUPER_API_KEY}" \
  -H 'content-type: application/json' \
  -d @- <<JSON
{
  "participantContextId": "${PARTICIPANT_CONTEXT_ID}",
  "did": "${PARTICIPANT_DID}",
  "active": true,
  "serviceEndpoint": {
    "id": "$(uuidgen)",
    "type": "CredentialService",
    "serviceEndpoint": "${WALLET_HOST}/api/credentials/v1/participants/${PARTICIPANT_CONTEXT_ID}"
  },
  "key": {
    "keyId": "${PARTICIPANT_DID}#key-1",
    "privateKeyAlias": "${PARTICIPANT_DID}-privatekey-alias",
    "keyGeneratorParams": { "algorithm": "EC" }
  }
}
JSON
```

Notes on the request body:

- `participantContextId` — a fresh UUID; it becomes the wallet-internal identifier and the path segment on `/v1/participants/{id}`. It is **not** the DID.
- `serviceEndpoint.type` must be `CredentialService`; this is the endpoint the issuer will POST issued VCs to, so the URL must be reachable from the issuer.

> - `serviceEndpoint.serviceEndpoint` must include `/v1/participants/${PARTICIPANT_CONTEXT_ID}` — the wallet consumes this value verbatim in the DID document, and the issuer follows it as-is when delivering credentials. A bare host produces a DID document that no issuer can honour.

Reviewer: shouldn't it be `/v1beta/` ?

Author: It's `/v1` intentionally — only the Identity *admin* API moved to `/v1beta` in 0.18. The Credential Service endpoint in the serviceEndpoint is the DCP protocol API, versioned independently: upstream `StorageApiController` is annotated `@Path("/v1/participants/{participantContextId}/credentials")` at v0.18.0. I also verified against our test wallet: `/api/credentials/v1/...` responds (401 auth required) while `/api/credentials/v1beta/...` is 404. Added a note in the doc since this is easy to trip over.

- `key.privateKeyAlias` — the Vault alias the wallet will use to store the generated private key. Any string is accepted; align it with your Vault path conventions.
- `key.keyGeneratorParams.algorithm` — `EC` (P-256) for interoperability with the current issuer profile. `EdDSA` / `Ed25519` also works but is not yet used across the MDS stack.

The response body includes the participant's **client id** and **client secret** — persist both:

```json
{
  "clientId": "did:web:my-connector.example.com",
  "clientSecret": "..."
}
```

The wallet will not return the client secret again. Store it in the **connector's** Vault under the alias the connector expects (`edc.iam.sts.oauth.client.secret.alias`; by convention `<did>-sts-client-secret`).

### Verify the DID document

```bash

> curl -fsS "https://<connector-host>/.well-known/did.json" | jq

Reviewer: ```suggestion
curl -fsS "https://<wallet-host>/.well-known/did.json" | jq
```

Author: Applied — the verify step now curls the wallet host, and the ingress note was updated to match.

```

You should see a document with `id = ${PARTICIPANT_DID}`, one `verificationMethod` entry (public key), and one `service` entry of type `CredentialService`.

If the connector's hostname is different from the wallet's hostname (typical), the connector's ingress must forward `/.well-known/did.json` to the wallet's `did` context. Any reverse proxy path rewrite that hides `/.well-known/` from the wallet breaks DID resolution.

### Request credential issuance

Once the connector's participant is registered on the wallet and on the issuer, ask the wallet to fetch a Membership credential from the issuer:

```bash
HOLDER_PID="$(uuidgen)"

curl -fsS -X POST \

>   "${WALLET_HOST}/api/identity/v1beta/participants/${PARTICIPANT_CONTEXT_ID}/credentials/request" \

Reviewer: ```suggestion
 "${WALLET_HOST}/api/identity/v1beta/participants/${url_encoded(PARTICIPANT_DID)}/credentials/request" \
```

I understood that with v1beta the API uses the url_encoded did:web instead of the old base64(did:web) , right?

Author: Correct — v1beta takes the URL-encoded participantContextId in the path. Strictly the path parameter is the contextId, not the DID, but I've switched the doc to use the DID *as* the contextId (same convention as the mds-identity-issuer seed scripts), so the example now URL-encodes it exactly as you suggested. (Side note from testing against our 0.17 wallet: v1alpha there actually accepts the *raw* contextId in the path — base64 returns 404 — so the encoding convention differs per version, one more reason to spell it out in the doc.)

  -H "x-api-key: ${PARTICIPANT_API_KEY}" \
  -H 'content-type: application/json' \
  -d @- <<JSON
{
  "holderPid": "${HOLDER_PID}",
  "issuerDid": "${ISSUER_DID}",
  "credentials": [
    {
      "id": "membershipCredential-id",
      "type": "MembershipCredential",
      "format": "VC1_0_JWT"
    }
  ]
}
JSON
```

> `PARTICIPANT_API_KEY` is the api-key returned in the participant-registration response (or the super-user key, which also works). Poll `GET /api/identity/v1beta/participants/${PARTICIPANT_CONTEXT_ID}/credentials/request/${HOLDER_PID}` until `status == "ISSUED"`.

Reviewer: Same for Participant Context ID

Author: Applied — same treatment as above: the contextId is the DID and every path uses the URL-encoded form (`PARTICIPANT_CONTEXT_ID_PATH`).

The example below assumes:

- `WALLET_HOST=https://wallet.example.com` — the wallet's public hostname.
- `PARTICIPANT_DID=did:web:wallet.example.com` — the DID your connector will advertise on DSP. It resolves on the **wallet's** hostname (the wallet serves the DID document), even though the connector is a different system on a different host.
- `SUPER_API_KEY` — the value of `EDC_IDENTITYHUB_SUPERUSER_API_KEY`.

> The DID doubles as the `participantContextId` — same convention as the [mds-identity-issuer seed scripts](https://github.com/Mobility-Data-Space/mds-identity-issuer/tree/main/scripts). Wherever the contextId appears in a URL path it must be URL-encoded (`did%3Aweb%3Awallet.example.com`):

Reviewer: The mds-identity-issuer is not a public repo. We can remove the note for now.
